WalletWasabi/Blockchain/Transactions/SmartTransaction.cs

```diff
@@ -83,6 +92,54 @@ public IReadOnlyCollection<IndexedTxOut> ForeignOutputs
 		}
 	}
 
+	/// <summary>Wallet inputs with the same script are virtually considered to be the same by blockchain analysis.</summary>
+	public IReadOnlyCollection<WalletVirtualInput> WalletVirtualInputs
+	{
+		get
+		{
+			if (WalletVirtualInputsCache is null)
+			{
+				WalletVirtualInputsCache = WalletInputs
+					.GroupBy(i => i.HdPubKey)
+					.Select(g => new WalletVirtualInput(g.Key, g.ToHashSet()))
+					.ToHashSet();
+			}
+			return WalletVirtualInputsCache;
+		}
+	}
+
+	/// <summary>Wallet outputs with the same script are virtually considered to be the same by blockchain analysis.</summary>
+	public IReadOnlyCollection<WalletVirtualOutput> WalletVirtualOutputs
+	{
+		get
+		{
+			if (WalletVirtualOutputsCache is null)
+			{
+				WalletVirtualOutputsCache = WalletOutputs
+					.GroupBy(o => o.HdPubKey)
+					.Select(g => new WalletVirtualOutput(g.Key, g.Sum(o => o.Amount), g.Select(o => new OutPoint(GetHash(), o.Index)).ToHashSet()))
+					.ToHashSet();
+			}
+			return WalletVirtualOutputsCache;
+		}
+	}
+
+	/// <summary>Foreign outputs with the same script are virtually considered to be the same by blockchain analysis.</summary>
+	public IReadOnlyCollection<ForeignVirtualOutput> ForeignVirtualOutputs
+	{
+		get
+		{
+			if (ForeignVirtualOutputsCache is null)
+			{
+				ForeignVirtualOutputsCache = ForeignOutputs
+					.GroupBy(o => o.TxOut.ScriptPubKey)
```

Comment:
There is a little problem when you compare two outputs based on their script pubkeys.

Imagine you have two outputs with script pubkeys derived from the same private key but one script pubkey is P2WPKH while the other is P2PKH. You consider these two outputs different. Nevertheless, the pubkey hash is the same in both the scripts. So it is obvious from the script pubkeys that these two outputs belong to the same person.

By the way, the same problem occurs between P2PKH (or P2WPKH) outputs and deprecated P2PK outputs.

Reply:
Makes sense now https://github.com/zkSNACKs/WalletWasabi/pull/8689/commits/3018eeaa8d71fe0b431e7bc244c8cd6f55f58a9f

Regarding HdPubKey, I think there was a bug in the original PR. The original PR did `hdPubKey.PubKey.ToBytes();`, but I think we should do `hdPubKey.PubKey.PubKeyHash.ToBytes()` as that's the equivalent for `ExtractScriptPubKeyParameters`:

https://github.com/zkSNACKs/WalletWasabi/pull/8689/commits/532f5963d977729777537a6f88a5d76563c71c0c

Reply:
> Regarding HdPubKey, I think there was a bug in the original PR. The original PR did hdPubKey.PubKey.ToBytes();, but I think we should do hdPubKey.PubKey.PubKeyHash.ToBytes() as that's the equivalent for ExtractScriptPubKeyParameters:

I can't remember if it was my intention or not but it certainly has no effect on the outcome and since we don't compare wallet virtual inputs (or outputs) with foreign virtual inputs (or outputs) the hashing of the key is unnecessary.